bmx160: Add support for BMX160 IMU Sensor

ssekar15 opened the pull request:
BMX160 IMU sensor is almost similar to BMI160 expect mangentometer bmm150 is built-in, so reuse existing driver with minimal changes. Now bmx160 can exist together with bmi160.

- Changes includes bmx160 accessed over I2C. 
- Added support for magnetometer and single & double tap events

ssekar15 committed:
drivers: sensors: bmx160: Add support for BMX160 IMU Sensor

BMX160 IMU sensor is almost similar to BMI160 expect mangentometer
bmm150 is built-in, so reuse existing driver with minimal changes.
Now bmx160 can exist together with bmi160.

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
samples: sensors: bmx160: add sample to test accel & gyro

Add a sample to demonstrate and test the bosch bmx160 imu sensor
sensor driver. Sample code covers accel, gyro, magn and tap function

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

gmarull commented on drivers/sensor/bmi160/bmx160_i2c.c:
```diff
@@ -0,0 +1,67 @@
+
+DEVICE_AND_API_INIT(bmx160, DT_INST_LABEL(0), bmx160_init, &bmx160_data,
+		&bmx160_config, POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,
+		&bmx160_api);
```
Driver should support an arbitrary number of instances, see `DT_INST_FOREACH` examples.

gmarull commented on drivers/sensor/bmi160/bmi160.h:
```diff
@@ -376,6 +425,50 @@ enum bmi160_odr {
 #	define BMI160_DEFAULT_ODR_GYR		13
 #endif
 
+#if defined(CONFIG_BMI160_MAG_ODR_RUNTIME) ||\
```
DeviceTree bindings options could be used instead of Kconfig options for defaults.

ssekar15 replied:
@gmarull agree for new driver, here just followed same way existing bmi160.

ssekar15 commented:
@MaureenHelm

gmarull commented on drivers/sensor/bmi160/bmi160_spi.c:
```diff
@@ -0,0 +1,71 @@
+
+	ret =  bm160_device_init(dev);
+	if (ret)
+		LOG_DBG("device init fail: %d", ret);
```
`{}` should be used on all if/else. ref. https://docs.zephyrproject.org/latest/contribute/index.html#coding-style

gmarull commented on drivers/sensor/bmi160/bmi160_spi.c:
```diff
@@ -0,0 +1,71 @@
+LOG_MODULE_REGISTER(BMI160, CONFIG_SENSOR_LOG_LEVEL);
+
+struct bmi160_device_data bmi160_data = {
+	.chipid = BMI160_CHIP_ID,
```
This is const data, should likely go to config.

gmarull commented on drivers/sensor/bmi160/bmi160_spi.c:
```diff
@@ -0,0 +1,71 @@
+
+#include "bmi160.h"
+
+#define BMI160_CHIP_ID		0xD1
```
Document constants, is this a device address? Is it always constant?

ssekar15 replied:
Its no device address, Deivce ID(kind of fuse value in register) is always constant based on BMX or BMI chip.

gmarull replied:
So then it should be defined where other constants are defined (bmi160.h). As it's a constant value for all devices, it can be used directly in the code.

ssekar15 replied:
I meant 0xD1 is for all the BMI160 devices and 0xD8 is for all the BMX160 devices, so cannot directly used in code.

gmarull replied:
I mean that as it's not instance specific, there's no need to store it as part of the driver config structure. Similar to Kconfig values.

ssekar15 replied:
Unless storing in driver config structure we cannot handle if both BMI160 and BMX160 device  used simultaneous in a board. Please correct me if I misunderstood.

gmarull replied:
Ok, I see. Thanks for the clarification.

gmarull commented on drivers/sensor/bmi160/bmi160_spi.c:
```diff
@@ -0,0 +1,71 @@
+	}
+
+	bmi160->spi_cfg.operation = SPI_WORD_SET(8);
+	bmi160->spi_cfg.frequency = DT_INST_PROP(0, spi_max_frequency);
```
Avoid instance-specific code in drivers. This makes impossible to have >1 sensor defined in DT

gmarull commented on drivers/sensor/bmi160/bmi160_spi.c:
```diff
@@ -0,0 +1,71 @@
+
+DEVICE_AND_API_INIT(bmi160, DT_INST_LABEL(0), bmi160_init, &bmi160_data,
+		&bmi160_config, POST_KERNEL, CONFIG_SENSOR_INIT_PRIORITY,
+		&bmi160_api);
```
Use a generic initializer, then `DT_INST_FOREACH_STATUS_OKAY`

gmarull commented on drivers/sensor/bmi160/bmi160_trigger.c:
```diff
@@ -183,17 +234,93 @@ static int bmi160_trigger_anym_set(struct device *dev,
+	int i;
+
+	if (attr == SENSOR_ATTR_LOWER_THRESH) {
+		if (val->val1 == 20) {
```
Constants should be defined and documented

gmarull commented on drivers/sensor/bmi160/bmi160.h:
```diff
@@ -427,9 +533,32 @@ struct bmi160_scale {
+	const char *bus_name;
+	u16_t i2c_addr;
 	struct spi_config spi_cfg;
+#if defined(DT_INST_0_BOSCH_BMX160_CS_GPIOS)
```
These macros are deprecated, you should use `devicetree.h` API.

gmarull commented on drivers/sensor/bmi160/bmi160_spi.c:
```diff
@@ -0,0 +1,80 @@
+	}
+
+	bmi160->spi_cfg.operation = SPI_WORD_SET(8);
+	bmi160->spi_cfg.frequency = DT_INST_PROP(0, spi_max_frequency);
```
This makes impossible to have >1 instance defined in DT. This driver already uses a generic initializer, so it can be stored in the instance config structure.

ssekar15 replied:
oops, my mistake

gmarull commented on drivers/sensor/bmi160/bmi160_trigger.c:
```diff
@@ -56,7 +81,10 @@ static void bmi160_handle_interrupts(void *arg)
 	union {
 		u8_t raw[6];
 		struct {
+/* dummy byte in beginning!! not working for I2C, kept to not break exising */
```
This problem should be described in more detail, referenced if possible.

ssekar15 replied:
I don't understand the background of dummy byte. my guess that used for struct padding, if so it should not be at beginning of structure. I don't have bmi160 device to confirm so.

gmarull commented on drivers/sensor/bmi160/bmi160_trigger.c:
```diff
@@ -183,17 +234,93 @@ static int bmi160_trigger_anym_set(struct device *dev,
+		rval = is_single ? BMI160_INT_S_TAP_EN : BMI160_INT_D_TAP_EN;
+	}
+
+	if (bmi160_reg_update(dev, BMI160_REG_INT_EN0, mask, rval) < 0) {
```
In general, it is better to propagate error codes (`r = call(); if (r < 0) return r;`)

gmarull commented on samples/sensor/bmx160/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
+include($ENV{ZEPHYR_BASE}/cmake/app/boilerplate.cmake NO_POLICY_SCOPE)
```
Should be `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`

gmarull commented on drivers/sensor/bmi160/bmx160_i2c.c:
```diff
@@ -0,0 +1,71 @@
+
+	ret =  bm160_device_init(dev);
+	if (ret)
+		LOG_ERR("device init fail: %d", ret);
```
missing braces

ssekar15 replied:
Not sure, why checkpatch not reported.

gmarull commented on drivers/sensor/bmi160/bmx160_i2c.c:
```diff
@@ -0,0 +1,71 @@
+#include <logging/log.h>
+#include "bmi160.h"
+
+#define BMX160_CHIP_ID		0xD8 /* Constant for bmx160 sensor */
```
Should likely be moved to `bmi160.h`

gmarull commented on drivers/sensor/bmi160/Kconfig:
```diff
@@ -208,4 +216,88 @@ config BMI160_GYRO_ODR_3200
+config BMI160_MAG_ODR_RUNTIME
+	bool "Set at runtime."
+
+config BMI160_MAG_ODR_25_32
```
Magnetometer is a BMX160 feature, so these Kconfig names are misleading.

ssekar15 replied:
Ok I try as much as possible to name it properly for folders, files, configs and magnetometer related things.

gmarull reviewed (changes requested):
BMI160 and BMX160 are two different sensors. Even though they are register compatible on common things, the driver should make this distinction clear. For example, keeping driver folder name `bmi160` is confusing.

gmarull commented on drivers/sensor/bm160/Kconfig:
```diff
@@ -3,14 +3,22 @@
+	depends on I2C
+	help
+	  Enable Bosch BMX160 inertial measurement unit that provides acceleration
+	  and angular rate measurements.
```
also provides magnetic field measurements

gmarull commented on drivers/sensor/bm160/CMakeLists.txt:
```diff
@@ -0,0 +1,11 @@
+zephyr_library_sources_ifdef(CONFIG_BMX160 bm160.c)
+zephyr_library_sources_ifdef(CONFIG_BMX160 bmx160_i2c.c)
+
+zephyr_library_sources_ifdef(CONFIG_BMI160_TRIGGER bm160_trigger.c)
```
trigger is not a `I` exclusive option

gmarull commented on drivers/sensor/bm160/Kconfig:
```diff
@@ -208,4 +216,88 @@ config BMI160_GYRO_ODR_3200
 endchoice
 
-endif # BMI160
+config BMX160_MAG
```
This option is not available for `I`, but it will still be shown on it.

ssekar15 replied:
No, config has depends on BMX160 so it will not shown to BMI160

gmarull commented on drivers/sensor/bm160/bm160.h:
```diff
@@ -376,6 +431,50 @@ enum bmi160_odr {
 
+#if defined(CONFIG_BMX160_MAG_ODR_RUNTIME) ||\
+		defined(CONFIG_BMX160_MAG_ODR_800)
+#	define BMI160_DEFAULT_ODR_MAG		11
```
mix of `BMI` and `BMX`

ssekar15 replied:
ok, will change it

gmarull commented on drivers/sensor/bm160/bm160.h:
```diff
@@ -384,11 +483,10 @@ struct bmi160_range {
 	gpio_pin_t int_pin;
 	gpio_dt_flags_t int_flags;
-#endif
+	u8_t chipid;
```
Why is the trigger ifdef removed?

ssekar15 replied:
Hard to define conditional compilation in macro expansion 

#define BMI160_INIT(idx)						\
									\
static const struct bmi160_device_config bmi160_config_##idx = {	\
	.chipid = BMI160_CHIP_ID,					\
	.gpio_port = DT_INST_GPIO_LABEL(idx, int_gpios),		\
	.int_pin = DT_INST_GPIO_PIN(idx, int_gpios),			\
	.int_flags = DT_INST_GPIO_FLAGS(idx, int_gpios),		\
};

gmarull commented on drivers/sensor/bm160/bm160.h:
```diff
@@ -401,18 +499,30 @@ union bmi160_pmu_status {
 #endif
 
-#define BMI160_BUF_SIZE			(BMI160_SAMPLE_SIZE)
+#define BMI160_BUF_SIZE		(BMI160_SAMPLE_SIZE + BMX160_MAG_SAMPLE_SIZE)
```
Why `I` size has relation with magnetometer?

ssekar15 replied:
Changes all the magnetometer related things to BMX but existing code will remains same. The idea is to make minimal changes to existing code to support for BMX magnetometer, consider change of common naming will leads to touch almost 70% of code.

Please share you feedback

gmarull reviewed (dismissed):
There are a few `I`/`X` inconsistencies that need to be solved.

MaureenHelm commented:
> BMI160 and BMX160 are two different sensors. Even though they are register compatible on common things, the driver should make this distinction clear. For example, keeping driver folder name `bmi160` is confusing.

Renaming existing Kconfigs and device tree compatibles is going to break out-of-tree usage of this driver. We also have precedent of sensor drivers like fxos8700 supporting additional register-compatible devices like mma84510 and mma8652. I'm not convinced we should rename bmi160. Can we instead enable the different functionality by adding an optional device tree property?

gmarull commented:
@MaureenHelm fair point regarding backwards compatibility. I think current approach is ok, `BMX160` options (i.e. magnetometer) use `BMX` prefix in Kconfig.

gmarull commented on drivers/sensor/bm160/Kconfig:
```diff
@@ -208,4 +216,88 @@ config BMI160_GYRO_ODR_3200
+
+endchoice
+
+endif # BMI160 || BMX160
```
magnetometer options apply to `BMI` when they should not

ssekar15 replied:
Sorry, statement not clear for me.

All magnetometer choice will be shown only when BMX160 is selected, achieved using below 
"depends on BMX160"
"depends on BMX160_MAG"

gmarull replied:
got the depends, I missed that point sorry!

gmarull reviewed (commented):
I have no more requests. As @MaureenHelm suggested, existing options (e.g. BMI160_TRIGGER) may be left with the same name for backwards compatibility. It would also be good to test these changes with a BMI160 sensor to make sure there are no regressions.

ssekar15 commented:
@gmarull preserved old configs untouched.
I don't have bmi160 device to test.

ssekar15 commented:
@gmarull @MaureenHelm
Ping for review/merge

ssekar15 commented:
re-based to latest.
@MaureenHelm I would suggest to merge if you have no more review comments, as @gmarull has satisfied and no more request.

siltemc commented:
What happened to the BMX160 SPI files? It is only i2c now. This no longer works with BMX160 sensor on an SPI bus.

ssekar15 commented:
@siltemc Could you tell what is broken ?

siltemc commented:
The BMX160 is both a I2C and SPI device like the BMI160. But your kconifg only allows the BMX160 to use I2C. There is also only a bmx160_i2c.c not bmx160_spi.c file. The sample doesn't need to be changed or anything just the driver should support SPI things. 
Also the SPI is missing the Chip Select pins/code in the bmi160 files. If the devices is on its own bus by itself, it would work fine I think. 
But the application I use this device in is on a shared bus with other SPI devices and which requires control of the CS pin.

ssekar15 commented:
> The BMX160 is both a I2C and SPI device like the BMI160. But your kconifg only allows the BMX160 to use I2C. There is also only a bmx160_i2c.c not bmx160_spi.c file. The sample doesn't need to be changed or anything just the driver should support SPI things.
This PR is mainly intended to add support for BMX160 support over I2C, SPI support shall be later or so. Nevertheless this PR doesn't introduce any regression or broken.

> Also the SPI is missing the Chip Select pins/code in the bmi160 files. If the devices is on its own bus by itself, it would work fine I think.
The actual driver of bmi160 doesn't have SPI chip select gpio change, also it work straight away for STM disco board. If I'm not wrong  you are testing on nrf board which needs SPI gpio configuration apart from specifying in dts (so not generic to handle in driver).

MaureenHelm reviewed (changes requested):
Please rebase to the latest master and fix the merge conflicts

ssekar15 committed:
drivers: sensor: bmi160: add support for magnetometer

add support to access bmx160 inbuilt magnetometer sensor output

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 committed:
drivers: sensor: bmi160: add support for tap events

add support to configure and event handling for single and double tap
using trigger

Signed-off-by: Saravanan Sekar <saravanan@linumiz.com>

ssekar15 commented on drivers/sensor/bm160/bm160.c:
```diff
@@ -452,6 +454,9 @@ static int bmi160_acc_config(struct device *dev, enum sensor_channel chan,
 		return bmi160_acc_slope_config(dev, attr, val);
+	case SENSOR_ATTR_LOWER_THRESH:
+	case SENSOR_ATTR_UPPER_THRESH:
+		return bmi160_acc_tap_config(dev, attr, val);
```
done

ssekar15 commented on drivers/sensor/bm160/bm160.c:
```diff
@@ -782,31 +805,12 @@ static int bmi160_channel_get(struct device *dev,
-	bmi160->spi_cfg.frequency = DT_INST_PROP(0, spi_max_frequency);
-	bmi160->spi_cfg.slave = DT_INST_REG_ADDR(0);
+	u8_t val = 0U;
+	s32_t acc_range, gyr_range;
```
done

avigreen1978 commented:
ping for review/merge?

MaureenHelm commented:
> ping for review/merge?

Please reopen and rebase

avigreen1978 commented:
> @avigreen1978 @MaureenHelm I will try to rebase to main and re-open by next week when I find time. Thanks.
Thanks, I would appreciate it

avigreen1978 commented:
Any updates?

muhady commented:
@parthitce: Any updates?
